core/src/main/java/org/elasticsearch/transport/TcpTransport.java

```diff
@@ -1053,6 +1053,23 @@ private BytesReference buildMessage(long requestId, byte status, Version nodeVer
         return new CompositeBytesReference(header, messageBody, zeroCopyBuffer);
     }
 
+    protected abstract int length(Buffer buffer);
```

Comment:
I'm not sure. I think this is relatively simple, it still keeps the core logic inside `TcpTransport`, but it has the advantage that it avoids an `BytesReference` allocation on every transport message.